src/ol/render/canvas/Immediate.js

```diff
@@ -1041,19 +1042,29 @@ class CanvasImmediateRenderer extends VectorContext {
       this.image_ = null;
       return;
     }
+    let pixelRatio = this.pixelRatio_;
+    const imageScale = imageStyle.getScaleArray();
+    if (imageStyle instanceof IconStyle) {
+      this.imageScale_ = [
+        pixelRatio * imageScale[0],
+        pixelRatio * imageScale[1],
+      ];
+      pixelRatio = 1;
```

Comment:
Not sure if this is the correct fix. It would be batter if icons would be created with the pixel ratio of the destination context, then they would also appear crisper on high dpi devices.

Reply:
Yes, images with `color` applied are available at higher dpi.  The ImageBuilder uses the image style's `getPixelRatio()` method to determine what is available, so that can also be used here.

Reply:
Updated to use high dpi images where available.